src/Control/Effect/Cut.hs

```diff
@@ -48,53 +49,64 @@ call m = send (Call m ret)
 --   prop> run (runNonDet (runCut (cut *> empty <|> pure a))) == []
 cut :: (Alternative m, Carrier sig m, Member Cut sig) => m ()
 cut = pure () <|> cutfail
 
 
 -- | The result of a nondeterministic branch of a computation.
-data Branch a
-  = Prune
+data Branch m a
+  = Cut
   | None
-  | Some a
-  deriving (Eq, Functor, Ord, Show)
+  | Pure a
+  | Alt (m a) (m a)
+  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
 
--- | Case analysis for 'Branch', taking a value to use for 'Prune', a value to use for 'None', and a function to apply to the contents of 'Some'.
+-- | Case analysis for 'Branch', taking a value to use for 'Cut', a value to use for 'None', and a function to apply to the contents of 'Pure'.
 --
---   prop> branch Prune None Some a == a
---   prop> branch a b (applyFun f) Prune == a
---   prop> branch a b (applyFun f) None == b
---   prop> branch a b (applyFun f) (Some c) == applyFun f c
-branch :: a -> a -> (b -> a) -> Branch b -> a
-branch a _ _ Prune    = a
-branch _ a _ None     = a
-branch _ _ f (Some a) = f a
+--   prop> branch Cut None Pure Alt a == (a :: Branch [] a)
+--   prop> branch a b (applyFun f) (applyFun2 g) (Cut :: Branch [] a) == a
+--   prop> branch a b (applyFun f) (applyFun2 g) (None :: Branch [] a) == b
+--   prop> branch a b (applyFun f) (applyFun2 g) (Pure c :: Branch [] a) == applyFun f c
+--   prop> branch a b (applyFun f) (applyFun2 g) (Alt c d :: Branch [] a) == applyFun2 g c d
+branch :: a -> a -> (b -> a) -> (m b -> m b -> a) -> Branch m b -> a
+branch a _ _ _ Cut       = a
+branch _ a _ _ None      = a
+branch _ _ f _ (Pure a)  = f a
+branch _ _ _ f (Alt a b) = f a b
+
+-- | Interpret a 'Branch' into an underlying 'Alternative' context.
+runBranch :: Alternative m => Branch m a -> m a
+runBranch = branch empty empty pure (<|>)
 
 
 -- | Run a 'Cut' effect within an underlying 'Alternative' instance (typically 'Eff' carrying a 'NonDet' effect).
 --
 --   prop> run (runNonDetOnce (runCut (pure a))) == Just a
 runCut :: (Alternative m, Carrier sig m, Effect sig, Monad m) => Eff (CutC m) a -> m a
-runCut = (>>= branch empty empty pure) . runCutC . interpret
+runCut = (>>= runBranch) . runCutC . interpret
 
-newtype CutC m a = CutC { runCutC :: m (Branch a) }
+newtype CutC m a = CutC { runCutC :: m (Branch m a) }
 
 instance (Alternative m, Carrier sig m, Effect sig, Monad m) => Carrier (Cut :+: NonDet :+: sig) (CutC m) where
-  ret = CutC . ret . Some
+  ret = CutC . ret . Pure
   eff = CutC . handleSum (handleSum
-    (eff . handle (Some ()) (branch (ret Prune) (ret None) runCutC))
+    (eff . handle (Pure ()) (bindBranch (ret Cut) runCutC))
     (\case
       Empty    -> ret None
-      Choose k -> runCutC (k True) >>= branch (ret Prune) (runCutC (k False)) (\ a -> ret (Some a) <|> runCutC (k False))))
+      Choose k -> runCutC (k True) >>= branch (ret Cut) (runCutC (k False)) (\ a -> ret (Alt (ret a) (runCutC (k False) >>= runBranch))) (fmap ret . Alt)))
```

Comment:
We now handle `Choose` by constructing `Alt`s, deferring choices to the underlying `Alternative` instance without building exponentially many branches.